### PR TITLE
[Variant]: Implement `DataType::Union` support for `cast_to_variant` kernel

### LINKED ISSUE #8195: [Variant]: Implement `DataType::Union` support for `cast_to_variant` kernel
Part of #6736 

**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

We implemented `cast_to_variant` for some primitive types in
- #8044

Now we need to complete out type support for other Arrow types.

**Describe the solution you'd like**

Add support for the types named in the title as well.

**Describe alternatives you've considered**

1. Add the corresponding match arms in the `cast_to_variant` kernel:
https://github.com/apache/arrow-rs/blob/c25c5a74c1c70cee57558e81a66a2e44725a67a6/parquet-variant-compute/src/cast_to_variant.rs#L70

2. add tests (e.g. https://github.com/apache/arrow-rs/blob/c25c5a74c1c70cee57558e81a66a2e44725a67a6/parquet-variant-compute/src/cast_to_variant.rs#L125-L142)

### PR DESCRIPTION
**Pull request by liamzwbao**
# Which issue does this PR close?

- Closes #8195.

# Rationale for this change

# What changes are included in this PR?

Implement `DataType::Union` for `cast_to_variant`

# Are these changes tested?

Yes

# Are there any user-facing changes?

New cast type supported

### CODE REVIEW

File: parquet-variant-compute/src/cast_to_variant.rs [hunk 1]
```diff
@@ -579,6 +598,34 @@ fn process_run_end_encoded<R: RunEndIndexType>(
     Ok(())
 }
 
+/// Convert dictionary encoded arrays
+fn convert_dictionary_encoded(
```
Comment (liamzwbao) [inline parquet-variant-compute/src/cast_to_variant.rs:621]: This change just moves the code from the `Dictionary` branch into this helper function. Since all the impls use slightly different coding styles, we could do a follow-up PR to make them consistent once all the variant cast implementations are complete.

File: parquet-variant-compute/src/cast_to_variant.rs [hunk 2]
```diff
@@ -544,8 +530,41 @@ pub fn cast_to_variant(input: &dyn Array) -> Result<VariantArray, ArrowError> {
...
+fn convert_union(
+    fields: &UnionFields,
+    input: &dyn Array,
+    builder: &mut VariantArrayBuilder,
+) -> Result<(), ArrowError> {
+    let union_array = input.as_union();
+
+    // Convert each child array to variant arrays
+    let mut child_variant_arrays = HashMap::new();
+    for (type_id, _) in fields.iter() {
```
Comment (klion26) [inline parquet-variant-compute/src/cast_to_variant.rs:558]: Do we need to merge the two passes into one?

File: parquet-variant-compute/src/cast_to_variant.rs [hunk 3]
```diff
@@ -544,8 +530,41 @@ pub fn cast_to_variant(input: &dyn Array) -> Result<VariantArray, ArrowError> {
...
+    // Process each element in the union array
+    for i in 0..union_array.len() {
+        let type_id = union_array.type_id(i);
+        let value_offset = union_array.value_offset(i);
+
+        if let Some(child_variant_array) = child_variant_arrays.get(&type_id) {
+            let value = child_variant_array.value(value_offset);
+            builder.append_variant(value);
+        } else {
+            // This should not happen in a valid union, but handle gracefully
```
Comment (alamb) [inline parquet-variant-compute/src/cast_to_variant.rs:577]: 👍

File: parquet-variant-compute/src/cast_to_variant.rs (hunk 2 above)
Comment (alamb) [reply parquet-variant-compute/src/cast_to_variant.rs:558]: I don't understand what you are suggesting

File: parquet-variant-compute/src/cast_to_variant.rs [hunk 4]
```diff
@@ -1620,6 +1667,93 @@ mod tests {
         assert_eq!(obj4.get("age"), None);
     }
 
+    #[test]
+    fn test_cast_to_variant_union_sparse() {
```
Comment (alamb) [inline parquet-variant-compute/src/cast_to_variant.rs:1692]: could we please add a test for a UnionArray where the child element is null? So that the output VariantArray has a null as well?

**Review by alamb — COMMENTED** [review]
Thanks @liamzwbao  -- this looks great. The only thing I think it is missing is a test for nulls in the UnionArray.

Thank you @klion26 for the review

File: parquet-variant-compute/src/cast_to_variant.rs (hunk 2 above)
Comment (liamzwbao) [reply parquet-variant-compute/src/cast_to_variant.rs:558]: Do you mean using one loop instead of two? That way we will compute the child array of each type_id on demand. 

But IIUC, we will use all the child arrays anyway if it's a valid union, and I think precomute all the child arrays may benefit the lookup in the second loop.

Comment (klion26) [reply parquet-variant-compute/src/cast_to_variant.rs:558]: Ah, yes, we'll use the same child_variant_array many times. The current would have a better performance

**Comment by alamb** [comment]
🚀
